File: LeetCode/Study Plan/Practice Problems/Medium/Simplify Path/Go/simplify_path.go
package main

import (
	"fmt"
	"strings"
)

func simplify_path(path string) string {
	s := strings.Split(path, "/")
	n := len(s)
	s_r := []string{}
	for i := 0; i < n; i++ {
		if s[i] != "" && s[i] != "." {
			s_r = append(s_r, s[i])
		}
	}
	st := []string{}
	n = len(s_r)
	for i := 0; i < n; i++ {
		if s_r[i] == ".." {
			if len(st) > 0 {
				st = st[:len(st)-1]
			}
		} else {
			st = append(st, s_r[i])
		}
	}
	return "/" + strings.Join(st, "/")
}

func main() {
	fmt.Println(simplify_path("/a/./b/../../c/"))
}

File: LeetCode/Medium/Simplify Path/Java/simplifyPath.java
class Solution {
    public String simplifyPath(String path) {
        String[] str = path.split("/");
        ArrayDeque<String> res = new ArrayDeque<>();

        for (String s : str) {
            if (s.equals("..")) {
                if (res.size() > 0) res.removeLast();
            } else if (s.equals(".") || s.length() == 0) continue;
            else res.add(s);
        }
        
        StringBuilder sb = new StringBuilder();
        Iterator it = res.iterator();
        while(it.hasNext()) {
            String s = it.next().toString();
            if(s.length() > 0) {
                sb.append(s + "/");
            }
        }
        if (sb.length() > 1) 
            sb.deleteCharAt(sb.length()-1);
        return "/" + sb.toString();
    }
}
